libvast/vast/detail/notifying_stream_manager.hpp

```diff
@@ -13,28 +13,37 @@
 
 #pragma once
 
+#include "vast/atoms.hpp"
 #include "vast/query_options.hpp"
 
 #include <caf/default_downstream_manager.hpp>
 #include <caf/detail/stream_stage_driver_impl.hpp>
 #include <caf/detail/stream_stage_impl.hpp>
 #include <caf/policy/arg.hpp>
+#include <caf/typed_response_promise.hpp>
+
+#include <utility>
 
 namespace vast::detail {
 
-template <class State>
-void notify_listeners_if_clean(State& st, const caf::stream_manager& mgr) {
-  if (!st.flush_listeners.empty() && mgr.inbound_paths_idle()
-      && mgr.out().clean()) {
-    st.notify_flush_listeners();
-  }
+inline bool should_notify_flush_listeners(
+  const std::vector<caf::typed_response_promise<atom::flush>>& flush_promises,
+  const caf::stream_manager& manager) {
+  return !flush_promises.empty() && manager.inbound_paths_idle()
+         && manager.out().clean();
+}
+
+inline void notify_flush_listeners(
+  std::vector<caf::typed_response_promise<atom::flush>>& flush_promises) {
+  for (auto&& flush_promise : std::exchange(flush_promises, {}))
+    flush_promise.deliver(atom::flush_v);
```

Comment:
Doesn't it make the whole forwarding mechanism irrelevant if the stream stage itself already delivers the promises?

Reply:
The vector inside the shared ptr is empty as soon as the first active partition was created.